### ISSUE TITLE
Undo Bug

### ISSUE THREAD

**Issue by abn5x** [FocalPaint/FocalPaint#8]
Seems there is a bug with auto-saving, as I was drawing a line it auto-saved (even though it supposed to save when 3 sec of not painting) the when I wanted to undo, it "forgot" the last two stroke , and starts undoing only after the third last:

![IMG_0294](https://user-images.githubusercontent.com/31189812/101085476-c5dbd700-35af-11eb-90c3-d43e58f92aac.jpeg)

**Comment by briend** [comment]
Thanks, I definitely released that too soon.  I had it happen to me and I started wondering if I could actually just auto-save in the background without even showing the "progress dialog" or interrupt painting, which should fix the undo synchronization.

Then I had other thoughts like, what if you just open a painting that you had saved and wanted to experiment (but not save it!).  The autosave would really ruin your day unless you had backups.  So, perhaps auto-save should save to a whole different folder, like  <whatever>-autosave. 

So I think it should be something like this might be safest:   

1.  You Create or Open a Painting "XYZ"  (the big folder name)
2. Auto-Save every N minutes to "XYZ-autosave" (overwrites each time)
3. If you click Save button it saves to "XYZ" and deletes "XYZ-autosave"
4.  The next auto-save recreates a new "XYZ-autosave"

So what if FocalPaint crashes at some point after the autosave.  Next launch, what should the Open Painting list show?
Should it show "XYZ-autosave" as well as "XYZ"?  Maybe that would be safest instead of trying to be clever about auto-recovery prompts.  You'll be able to open either one and decide which one to keep.

Oh, I think if you open the XYZ-autosave copy it should be treated as a whole new folder name, so that autosave can continue working without ending up with "folder-autosave-autosave-autosave".  Definitely means more cleanup is required but I'd rather be safe than sorry.  I actually had FocalPaint totally eat my homework for a class I'm taking, due to an off-by-one index mistake. . . :-)

**Comment by briend** [comment]
I think v1.34 Build 2 should pretty much fix this.  I really didn't think it could autosave in the background and be unnoticed, but it seems that way!

**Comment by abn5x** [comment]
Hahahah it happens to me sometimes too, I think that the best thing would be to give the user all these options in “App Settings”, some may want the work to auto-save some may not, if the app crashes it would definitely be amazing to choose from the latest autosave and the last manual save. Also the auto-save frequency would be nice to be adjustable too.

Ofc the default should be with auto-save enabled, and some normal interval, but there should always be the option to change or disable those settings :)

**Comment by abn5x** [comment]
btw, this article explains some concepts of the AFPS FS which might interest you in the auto-save feature and saving in general

https://eclecticlight.co/2020/04/14/copy-move-and-clone-files-in-apfs-a-primer/

**Comment by briend** [comment]
Thanks @abn5x. I didn't realize there were efficient file cloning methods available for iOS.  That might be really cool for making duplicates of files.  I might need to rework a lot of stuff though, since I re-write the whole file every time you save it.  Not sure if that would un-clone it, so to speak.  I'll close this issue though since I think it's pretty fixed, albeit the file handling is still admittedly clunky.
